doc/developer/style.md

```diff
@@ -116,6 +116,77 @@ Some key principles to highlight are:
 * Avoid contractions and spell out words in full.
 * Word choices to be mindful of: find vs. exists, may vs. can vs. might.
 
+## Log message style
+
+We use the [`tracing` crate](https://docs.rs/tracing/latest/tracing/)'s log
+macros. Observe the following guidelines:
+
+  * Use sentence fragment style. Say "beginning operation", not
+    "Beginning operation."
+
+  * Prefer structured fields to string formatting:
+
+    ```
+    fn handle_persist_thingy(shard_id: ShardId) {
+        // Preferred.
+        info!(shard_id = ?the_shard, "handling persist thingy");
+        // Acceptable.
+        info!("handling persist thingy for shard {}", shard_id);
+    }
+    ```
+
+Messages at each level must meet the indicated standard:
+
+* **`ERROR`**: Reports an error that certainly indicates data corruption, data
+  loss, unavailability, or an invariant violation that results in undefined
+  behavior. These errors are reported to Sentry and may eventually page the
+  on-call engineer. Use *judiciously*.
+
+  Examples:
+
+  * A referenced collection is not present in the storage controller.
+  * A source ingested the deletion of a record that does not exist, causing a
+    "negative multiplicity."
+
+* **`WARN`**: Like `ERROR`, but when there is some uncertainty about the
+  error condition.
+
+  * A cancellation request was received for an unknown peek.
+
+* **`INFO`**: Reports normal system status changes. Messages at this level are
```

Comment:
the way persist uses INFO right now is roughly: things that you might want to know if you're actively investigating a certain issue, but you should 100% ignore otherwise

my current favorite example of INFO in persist right now is when a given request to an external system has failed the first N retries (N currently is 3)

Reply:
Gotcha. How does `persist` think about the difference between `DEBUG` and `INFO`?

Reply:
I think of it pretty much as a pure practicality tradeoff between "volume" and "likelihood of usefulness"

Reply:
Ack. Updated accordingly!

Reply:
@danhhz, @benesch  , Anecdotally I think the INFO logs from Persist are way too verbose to the extent that it causes non-trivial issues with debugging non-Persist issues. This isn't a criticism -- it mainly stems from the fact that Persist has much more comprehensive logging than other teams, which is a good thing. But I think a lot of what is there now could be downgraded to DEBUG. There have been several times where I was attempting to debug something that turned out to be non-Persist-related and got confused by the large volume of INFO statements from Persist crowding out everything else.

Reply:
a number of the spammier logs that have survived are still there because they've helped debug an actual issue (that I'm not sure how we'd have debugged another way). I'm always happy to revisit the tuning, though. any of them in particular?

also, is this local dev or prod? if local dev, I think there's a world in which we _somehow_ set a different log filter for `./bin/environmentd` than in prod (and tune mz_persist_client down to WARN). if prod, my personal take here is that we'll get better logs tooling over time, which will let us tune based on log infra resource usage instead of optimizing for someone trying to visually scan through the raw dump of everything